.github/bin/download-maven-dependencies.sh

```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+# Enable all known profiles to make sure their plugin dependencies are downloaded
+# GIB should be disabled even though it's profile is activated, to make sure it doesn't skip any submodules
```

Comment:
Why would we want to download dependencies for modules we're not going to build?
Isn't it wasteful?

Reply:
> Why would we want to download dependencies for modules we're not going to build? Isn't it wasteful?

I'm not sure if having GIB activated affects dependency download tools in any way, but I agreed with @nineinchnick to add this as a precaution.

Our current caching strategy is "if there was no cache hit, register a cleanup hook to upload the local repo into the GitHub cache". That's why other jobs remove their local repo entirely, so that only `maven-checks` uploads dependencies after removing any locally cached Trino artifacts.

If GIB were to remove any dependencies, another PR with identical POM files would download the incomplete cache, exposing itself to the risk of Maven Central failures in all jobs. The other PR would NOT update the cache, since it got a cache hit, so that risk would propagate forward until the cache entry is evicted.

Reply:
> If GIB were to remove any dependencies, another PR with identical POM files would download the incomplete cache

what's the cache key?

Reply:
> what's the cache key?

`${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}`

Reply:
I would also point out that I don't aim to be perfect here - it's OK to miss a dependency or two.
As long as the number of dependencies which have to be downloaded during build goes down, we're decreasing chances of a Maven Central failure taking our build down, and we can consider that an improvement.

Reply:
> > what's the cache key?
> 
> `${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}`

Thanks. Let's add some mention of that: `# Downloading dependencies is used to populate the maven cache shared between PRs, so PR-specific GIB state needs to be ignored`

> I would also point out that I don't aim to be perfect here - it's OK to miss a dependency or two.

I was under impression than one missing dependency was the cause of https://github.com/trinodb/trino/issues/15793.

obviously i am wrong, but i would very very much prefer a clean approach where (a) we download dependencies with all due retries and (b) we run the build, ideally with offline flag to validate (a).

doing this would solve https://github.com/trinodb/trino/issues/15794 too

Reply:
That one dependency was missing because it wasn't downloaded by `dependency:go-offline` as expected, and Maven Central happened to fail at the same time. I've manually checked that adding the second tool downloads the missing POM on my laptop, so it should fix #15793.

A clean approach with `--offline` would definitely be better, but I have yet to find a perfect dependency download tool.
If we decide that we're willing to risk it, we can disable retries on regular Maven builds and only leave it on for dependency downloads.